### PR TITLE
Add Buffer, CBOR, Request and Result contracts and libraries

### PR DESCRIPTION
**Pull request by aesedepece**
This adds 5 new contracts and libraries:
- **Buffer**: a convenient wrapper around the `bytes memory` type that exposes a buffer-like interface that you can read from, seek and rewind.
- **CBOR** + **CBORValue**: a decoder library for Concise Binary Object Representation (CBOR) data format (RFC7049), along an algebraic data type providing an ergonomic API for handling CBOR/RADON types as native Solidity types.
- **Request**: a pseudo-interface for Witnet data requests whose constructor automatically computes the ID for the request. The constructor is provided for the sake of consistency and trustlessness. If the ID was hard-coded in the contract, then it may easily differ from the real one, and thus the contract would be resoluble with the result of another request that may use different sources and whatnot. 
- **Result**: an algebraic data type built around a `CBORValue` in the like of Rust's `Result<T, E>`. Provides convenient methods to safely get the inner value or error without crashing the calling contract.

### CODE REVIEW

File: test/data.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+      "proof": [3, 189, 34, 13, 252, 163, 1, 216, 28, 52, 195, 252, 136, 198, 186, 174, 242, 214, 122, 94, 226, 69, 60, 186, 15, 53, 238, 154, 21, 168, 39, 83, 113, 206, 139, 114, 36, 223, 19, 115, 146, 146, 236, 3, 62, 172, 149, 230, 100, 171, 36, 58, 102, 158, 68, 199, 203, 58, 122, 39, 96, 85, 247, 88, 242, 136, 232, 68, 226, 229, 28, 141, 44, 110, 197, 162, 43, 133, 6, 247, 235]
+    },
+    {
+      "lastBeacon": [68, 222, 242, 94, 4, 103, 175, 47, 72, 236, 24, 132, 7, 222, 206, 177, 149, 206, 208, 44, 147, 112, 74, 180, 131, 144, 217, 178, 52, 91, 124, 107, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2],
+      "proof": [2, 81, 255, 187, 96, 151, 164, 77, 173, 238, 191, 134, 19, 6, 47, 110, 113, 188, 251, 130, 228, 84, 245, 208, 34, 111, 189, 190, 117, 167, 67, 240, 232, 34, 123, 249, 55, 13, 89, 218, 86, 172, 183, 113, 73, 118, 100, 53, 236, 30, 246, 82, 214, 193, 172, 160, 52, 219, 252, 66, 121, 51, 174, 131, 71, 244, 77, 216, 205, 194, 43, 219, 169, 79, 61, 90, 119, 13, 206, 30, 134]
+    }
+<<<<<<< HEAD
```
Comment (aesedepece) [inline test/data.json]: Ooops!

**Comment by aesedepece** [comment]
I'll call this ready for review.

File: contracts/Buffer.sol [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      (uint16(uint8(bytesValue[2])) << 8) |
+      uint8(bytesValue[3]);
+  }
+
+  /**
+  * @notice Read and consume the next 8 bytes from the buffer as an `uint64`.
+  * @param _buffer An instance of `BufferLib.Buffer`.
+  * @return The `uint64` value of the next 4 bytes in the buffer counting from the cursor position.
+  */
+  function readUint64(Buffer memory _buffer) internal pure returns (uint64) {
```
Comment (mariocao) [inline contracts/Buffer.sol]: As we saw together today, these functions to read from bytes are much more efficient in assembly. :blush:

**Comment by aesedepece** [comment]
Superseded by #24

**Pull request closed without merging**